### DISCUSSION TITLE
Pixmap.save Unable to save pNG format correctly

### DISCUSSION THREAD

**Discussion by jianghwjx** [pymupdf/PyMuPDF#2869] [Looking for help] [answered]
I used python to write a pdf format invoice into a png image program, found that the new version of PyMuPDF Pixmap.save () save the png format may appear to stamp opaque problem, try to use 1.20.0 before the version of the library, use Pixmap.writePNG () save is normal, I would like to ask about the new I would like to ask the new version there are other methods equivalent to Pixmap.writePNG command to achieve the same function?

**Comment by JorjMcKie** [comment]
You are referring to method names that have been changed a long time ago now.
Saving a pixmap to a file now simply is `pix.save(filename)`. Please note that JPEG is now among the image formats that are supported by PyMuPDF natively. So if the filename endswith ".jpg" or ".jpeg" JPEG format is used.

Saving to memory is now called `pix.tobytes("format")`, where format is any of the supported output format as above.

**Reply by jianghwjx** [reply]
I know that the command has been changed to pix.save(). The problem I described is that some images saved using pix.save(), whether they are PNG or JPG, cannot be made opaque from images that were originally transparent in PDFs
Like the green circle in the picture below
![image](https://github.com/pymupdf/PyMuPDF/assets/35679970/0de11828-5f0e-480b-97d5-447d2443340e)

**Reply by jianghwjx** [reply]
Using the original command to save PNG normally should look like this
![image](https://github.com/pymupdf/PyMuPDF/assets/35679970/3b983da9-c8bf-442c-a86f-77836cf13454)

**Reply by JorjMcKie** [reply]
The image with the red stamp obviously is not a transparent image obviously. So when it was created from a Pixmap that Pixmap did not have an alpha channel, pix.alpha = 0.
Why this was or was not the case depends on your upstream app creating the pixmap.
Otherwise, only PNG images can be transparent - JPGs never are.

**Reply by jianghwjx** [reply]
The PDF file uses this
[通行费.pdf](https://github.com/pymupdf/PyMuPDF/files/13667808/default.pdf)
The same PDF uses the following code, version 1.18.19, and saving the PNG is transparent
```
def to_png(input_file):
    name = input_file[:-4]
    name2 = input_file.split('\\')[-1]
    pdf = fitz.open(input_file)
    rname = ''

    for page_num in tqdm(range(pdf.page_count), desc=f'正在转换：[{name2}]'):
        pix = pdf[page_num].get_pixmap(matrix=fitz.Matrix(3, 3))
        rname = f"{name}_page_%i.png" % (page_num + 1)
        pix.writePNG(rname)
```
The new version is opaque when saved using pix.save
```
def to_png(input_file):
    name = input_file[:-4]
    name2 = input_file.split('\\')[-1]
    pdf = fitz.open(input_file)
    rname = ''

    for page_num in tqdm(range(pdf.page_count), desc=f'正在转换：[{name2}]'):
        pix = pdf[page_num].get_pixmap(matrix=fitz.Matrix(3, 3))
        rname = f"{name}_page_%i.png" % (page_num + 1)
        # pix.writePNG(rname)
        pix.save(rname)
```
Because our English is not very good, is it a problem with my understanding, or is it that pix.save does not support saving PNG format

**Reply by JorjMcKie** [reply]
> is it a problem with my understanding,

I think so, yes. Whether or not a pixmap has an alpha channel is defined at creation time. Please look at the documentation for `Page.get_pixmap()`. When you say `alpha=True` the pixmap will have an alpha channel. The default is `alpha=False`.
I don't remember whether the default under version 1.18.x was any different.

But saving a pixmap will not change transparency.

**Reply by jianghwjx** [reply]
First of all, thank you for your patient reply
Actually, I have already tried ```alpha=True``` before giving feedback on this issue. 
Below is the png for saving the execution results of version 1.18.19
```
    for page_num in tqdm(range(pdf.page_count), desc=f'正在转换：[{name2}]'):
        pix = pdf[page_num].get_pixmap(matrix=fitz.Matrix(3, 3), alpha=True)
        rname = f"{name}_page_%i.png" % (page_num + 1)
        pix.writePNG(rname)
        # pix.save(rname)
```
The saved image looks like this
![2通行费_page_1](https://github.com/pymupdf/PyMuPDF/assets/35679970/d46abf7c-62f7-4885-aea2-cfc3fa73b9d4)

However, using the new version, the code is as follows
```
    for page_num in tqdm(range(pdf.page_count), desc=f'正在转换：[{name2}]'):
        pix = pdf[page_num].get_pixmap(matrix=fitz.Matrix(3, 3), alpha=True)
        rname = f"{name}_page_%i.png" % (page_num + 1)
        # pix.writePNG(rname)
        pix.save(rname)
```
The saved image looks like this
![1通行费_page_1](https://github.com/pymupdf/PyMuPDF/assets/35679970/136c037a-dce6-4a36-a19f-6cae488fb015)

The above tests used the same PDF file
[通行费.pdf](https://github.com/pymupdf/PyMuPDF/files/13680008/default.pdf)

**Comment by JorjMcKie** [comment]
Cannot reproduce. The following snippet works exactly as it should:
```python
doc=fitz.open("default.pdf")
page=doc[0]
pix=page.get_pixmap(matrix=fitz.Matrix(3,3),alpha=True)
pix.save("test.png")
fitz.__version__
'1.23.7'
```
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/7cdbdd21-0c43-4183-a845-63212f4f3de9)
With alpha=False:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/6de4cd09-dc4e-4efa-bca9-43072e1e8635)

**Reply by jianghwjx** [reply]
The situation is indeed a bit strange, this is the result of my test
![image](https://github.com/pymupdf/PyMuPDF/assets/35679970/da47bc41-0f3e-40e8-92ba-f44632f95be2)
and
![image](https://github.com/pymupdf/PyMuPDF/assets/35679970/cb83db53-a8ba-4dbc-b273-c7434c442911)
Not sure what other information I can provide to locate the problem

**Reply by JorjMcKie** [reply, marked as answer]
~Well best would be everything that you haven't mentioned yet, notably operating system and PyMuPDF installation method.~

Found the problem:
It is connected with an issue in the base library MuPDF. My local PyMuPDF version already uses the current, corrected MuPDF, therefore I am able to produce correct images.

So you will have to wait with your upgrade until PyMuPDF v1.23.8 comes out.

**Reply by jianghwjx** [reply]
Thank you very much for your answer, you are really very patient, thank you again

**Reply by jianghwjx** [reply]
Has version 1.23.8 not fixed this issue yet 
After attempting to upgrade, this issue still exists

**Comment by jianghwjx** [comment]
这个问题是不是还没有在新版本中修正，因为我尝试使用1.23.8及以后的版本，依然存在这个问题，能确认一下吗
Has this issue not been fixed in the new version yet? I tried using versions 1.23.8 and later, but the issue still exists. Can you confirm

**Reply by JorjMcKie** [reply]
Yes, unfortunately.
The fix must happen in MuPDF, not in PyMuPDF. The MuPDF team is still working on this.

**Reply by jianghwjx** [reply]
Okay, then I'll keep paying attention. Thank you.

**Reply by yash-0501** [reply]
Is this issue, fixed yet? @JorjMcKie, I am facing the same kind of issue
